File: internal/cloudsync/retry.go
package cloudsync

import (
	"errors"
	"time"
)

// RetryDecision says whether an operation may be retried and with what initial
// backoff. Errors that require user action (auth, permission, quota, schema,
// invalid response, unsupported capability) never spin; precondition failures
// are reconciled by the engine rather than blindly retried.
type RetryDecision struct {
	Retryable bool
	Backoff   time.Duration
}

// ClassifyRetry maps a store/provider error onto a retry decision.
func ClassifyRetry(err error) RetryDecision {
	var se *StoreError
	if !errors.As(err, &se) {
		return RetryDecision{Retryable: false}
	}
	switch se.Kind {
	case ErrRateLimit:
		b := time.Second
		if se.RetryAfter > 0 {
			b = se.RetryAfter
		}
		return RetryDecision{Retryable: true, Backoff: b}
	case ErrRetryableTransport:
		return RetryDecision{Retryable: true, Backoff: time.Second}
	default:
		return RetryDecision{Retryable: false}
	}
}

File: frontend/src/sync/core/retry.ts
// Retry classification mirroring internal/cloudsync/retry.go. Errors that
// require user action (auth, permission, quota, schema, invalid response,
// unsupported capability) never spin; precondition failures are reconciled by
// the engine rather than blindly retried.

import { StoreError } from './remoteStore'

export interface RetryDecision {
  retryable: boolean
  backoffMs: number
}

/** Maps a store/provider error onto a retry decision. */
export function classifyRetry(err: unknown): RetryDecision {
  if (err instanceof StoreError) {
    switch (err.kind) {
      case 'rate-limit':
        return { retryable: true, backoffMs: err.retryAfterMs ?? 1000 }
      case 'retryable-transport':
        return { retryable: true, backoffMs: 1000 }
      default:
        return { retryable: false, backoffMs: 0 }
    }
  }
  return { retryable: false, backoffMs: 0 }
}
